src/foam/dao/RoutingJournal.js

```diff
@@ -38,68 +77,141 @@ try {
       `
     },
     {
-      name: 'writePut_',
-      javaThrows: [
-        'java.io.IOException'
-      ],
-      args: [
-        {
-          name: 'x',
-          type: 'Context'
-        },
-        {
-          name: 'record',
-          type: 'String'
-        },
-        {
-          name: 'c',
-          type: 'String'
+      name: 'put',
+      javaCode: `
+        this.put_(x, dest, null, nu);
+      `
+    },
+    {
+      name: 'put_',
+      javaCode: `
+        if ( ! getReplayed() ) return;
+        putWithPrefix_(x, old, nu, dest);
+      `
+    },
+    {
+      name: 'remove',
+      javaCode: `
+        if ( ! getReplayed() ) return;
+        removeWithPrefix_(x, obj, dest);
+      `
+    },
+    {
+      name: 'assertReplayDAORegistered_',
+      args: [ { name: 'service', type: 'String' } ],
+      javaCode: `
+        if ( ! getReplayDAOs().containsKey(service) ) {
+          throw new RuntimeException(String.format(
+            "Service '%s' completed initialization without registering a " +
+            "replay DAO", service));
         }
+      `
+    },
+    {
+      name: 'getReplayDAO_',
+      type: 'foam.dao.DAO',
+      args: [
+        { name: 'service', type: 'String' }
       ],
       javaCode: `
-        String service = (String) x.get("service");
-        write_(sb.get()
-          .append(service)
-          .append(".p(")
-          .append(record)
-          .append(")")
-          .toString());
+        if ( getReplayDAOs().containsKey(service) ) {
+          return getReplayDAOs().get(service);
+        }
+
+        getReplayDAOWaitObjectsLock().lock();
+
+        AtomicBoolean waitObj = null;
+
+        if ( getReplayDAOWaitObjects().containsKey(service) ) {
+          getReplayDAOWaitObjectsLock().unlock();
+          waitObj = getReplayDAOWaitObjects().get(service);
+          synchronized ( waitObj ) {
+            while ( ! waitObj.get() ) {
+              try {
+                waitObj.wait();
+              } catch ( InterruptedException e ) {
+                // Interrupt invokes retry
+                continue;
+              }
+            }
+          }
+          return getReplayDAO_(service);
+        }
+
+        waitObj = new AtomicBoolean();
+
+        synchronized ( waitObj ) {
+          getReplayDAOWaitObjects().put(service, waitObj);
+          getReplayDAOWaitObjectsLock().unlock();
```

Comment:
put is done inside this synchronized block to prevent waitObj.notifyAll() (inside registerReplayDAO method) from being called before waiting starts (which would result in a deadlock if it happens during the while condition below)